supports tcpping and tcpconn and add error handling of connection errors

chyeh opened the pull request:
**[OWL-611] Implementation of supporting `tcpping` & `tcpconn`**

It supports `tcpping` and `tcpconn` now. Probes are executed in corresponding goroutines now. The code was refactored and the unit tests was added. 

**[OWL-659] Not interrupted by errors on connections**

Fixed the problem of being interrupted by the unsuccessful connection to
both the RPC server `hbs` and the HTTP server `agent`.

chyeh committed:
[OWL-611] Support `tcpconn` as the probing utility

In this commit, the code is ugly. It needs for code refactorying using
dependency injection pattern.

chyeh committed:
[OWL-611] Corresponding changes for `go test`

chyeh committed:
[OWL-611] Refactor the code by dependency pattern

chyeh committed:
[OWL-611] Code refinement and refactoring

chyeh committed:
[OWL-611] Change the arguments of `Task()`

chyeh committed:
[OWL-611] Code refinement & correlative unit test

`measure.go`.`parse()` -> `parse.go`.`Parse()` + `parse_test.go`

chyeh committed:
[OWL-611] Refactor and refine the code

Add `calc.go` & `calc_test.go` for refactoring the part of the
statistics. In this commit, some dead code was removed and some
variables' names were changed.

chyeh committed:
Comment updated

chyeh committed:
[OWL-611] `data.go`->`marshal.go`

Renamed `data.go` as `marshal.go`. Refactoring and refinement of the
code for marshalling the JSON parameters were also done.

chyeh committed:
[OWL-611] Add unit tests for `marshal.go`

chyeh committed:
[OWL-611] Add unit tests for `parse.go`

chyeh committed:
[OWL-611] Add unit test for `marshalStatsRow()`

chyeh committed:
[OWL-611] Support `tcpping`

Because the metrics collected from `tcpping` are similar to the metrics
from `fping`, it can share the same functions with `fping`.

chyeh committed:
[OWL-611] Update some logging information

chyeh commented:
I just found that different probes are using the `FpingInterval` field as their invervals

chyeh committed:
[OWL-611] Update an error logging info

chyeh committed:
[OWL-611] Make probes run by respective intervals

hitripod commented on measure.go:
```diff
@@ -2,49 +2,163 @@ package main
-	go fping()
-	go tcpping()
-	go tcpconn()
+	go func() {
```
Could the agent collect the metrics using only `fping`?

chyeh replied:
How about adding some fields in the `cfg.json`, which let you enable/disable different types of probes?

hitripod replied:
It should be controlled by HBS not `cfg.json`, please refer to [open-falcon agent](https://github.com/Cepave/agent/blob/develop/cron/collector.go#L28).

```
    for _, v := range funcs.Mappers {
        go collect(int64(v.Interval), v.Fs)
    }
```

chyeh committed:
[OWL-611] Code refactoring and refinement

chyeh committed:
[OWL-659] Not interrupted by errors on connections

Fixed the problem of being interrupted by the unsuccessful connection to
both the RPC server `hbs` and the HTTP server `agent`.

chyeh committed:
Remove `init()` in `query_test.go`

Sometimes it conflicts with onther test files.

chyeh committed:
[OWL-611] Enable/Disable the probes by `hbs`.

It checks the `Command` field in the response from `hbs`. The `Command`
field tells what kind of utilities should be used for the measurements.

chyeh committed:
[OWL-611] Code refactory and tests in `Query()`

chyeh committed:
[OWL-663] Copy `hbsResp` in the measurements

It doesn't access the `GetGeneralConfig.hbsResp` in the measurements
now. Instead, it copies the necessary data at the beginning of the
measurements, which is in `Task()`.

chyeh committed:
[OWL-663] Variable's name: `lock` -> `jsonCfgLook`

chyeh committed:
[OWL-663] Code refinement

chyeh committed:
[OWL-663] `hbsResp` not accessed after copied

It was a mistake that `ProbingCommand()` still accessed `hbsResp`. It
has been fixed.

chyeh committed:
[OWL-663] `atomic.Value` for concurrency

I made `GeneralConfig.hbsResp` a `Value` variable. The correlative unit
test in `task_test.go` was updated to.
